### PR TITLE
ASAN build for ray core tests

### PR DESCRIPTION
**Pull request by ericl**
This PR adds an ASAN bazel profile and adds a Travis build that runs Ray core unit tests with ASAN enabled.

To debug ASAN errors locally, you must first set the right LD_PRELOAD for libasan in .bazelrc, and then you can simply run: `bazel test --config=asan --build_tests_only //:all`.

Unfortunately, ASAN adds too much memory overhead to run for all Ray Python tests. We can revisit enabling it if we get higher memory build machines.

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 1]
```diff
@@ -386,8 +386,8 @@ class SchedulingQueue {
     while (!pending_tasks_.empty()) {
       auto head = pending_tasks_.begin();
       head->second.Cancel();
-      pending_tasks_.erase(head);
       next_seq_no_ = std::max(next_seq_no_, head->first + 1);
+      pending_tasks_.erase(head);
```
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.h:390]: Use after iterator invalidation caught by asan.

File: src/ray/common/scheduling/scheduling_test.cc [hunk 2]
```diff
@@ -640,7 +640,7 @@ TEST_F(SchedulingTest, TaskResourceInstancesTest) {
 
     TaskRequest task_req;
     vector<FixedPoint> pred_demands = {3. /* CPU */, 2. /* MEM */, 1.5 /* GPU */};
-    vector<bool> pred_soft = {false};
+    vector<bool> pred_soft = {false, false, false};
```
Comment (ericl) [inline src/ray/common/scheduling/scheduling_test.cc:643]: Interestingly ASAN was not able to detect these out of bounds array accesses, but the test behavior changed enough that it failed.

**Review by edoakes — APPROVED** [review]
Awesome! Is it possible to print the instructions for reproducing if the ASAN build fails? Or at least a pointer to the instructions in `.bazelrc`? That would probably save us some time in the future with people asking how to repro locally.

**Comment by ericl** [comment]
Moved the ld preload flag into the travis command line, maybe that will make it easier to figure out what needs to be changed.

**Comment by ericl** [comment]
Oops, that seems to have done something bad, reverting.

**Comment by ericl** [comment]
Going to merge for now, we can revisit how to make the error more obvious when it comes up.

**Pull request merged**